### PR TITLE
Tons of stuff

### PR DESCRIPTION
**Pull request by pitust**
A lot of stuff, mainly strinigification and proto, unfortunatly squashed into 4 commits. Sorry for my bad organization.

### CODE REVIEW

**Comment by pitust** [comment]
Now that I look, there are tons of oversights.

File: native-vm-just-run-ninja/.gdb_history [hunk 1]
```diff
@@ -0,0 +1,12 @@
+file ./out
```
Comment (coder-mike) [inline native-vm-just-run-ninja/.gdb_history:1]: Not sure all these files should be committed to git.

File: native-vm-just-run-ninja/build.ninja [hunk 2]
```diff
@@ -0,0 +1,19 @@
+cxx = g++
```
Comment (coder-mike) [inline native-vm-just-run-ninja/build.ninja:1]: Can you give me an overview of how ninja is integrated into the build or test pipeline?

File: native-vm-just-run-ninja/project.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
+#include <iostream>
```
Comment (coder-mike) [inline native-vm-just-run-ninja/project.cpp:1]: What is this file for?

File: test/getting-started/code/microvium/microvium.c [hunk 4]
```diff
@@ -4761,9 +4762,22 @@ static TeError getProperty(VM* vm, Value objectValue, Value vPropertyName, Value
...
+      Value proto;
+      TeError e = getProperty(vm, objectValue, VM_VALUE_STR_PROTO, &proto);
+      if (e != MVM_E_SUCCESS) return e;
+      if (proto == VM_VALUE_NULL) {
+        CODE_COVERAGE(600); // Hit
+        *vPropertyValue = VM_VALUE_UNDEFINED;
+        return MVM_E_SUCCESS;
+      }
+      CODE_COVERAGE(601); // Hit
+      return getProperty(vm, proto, vPropertyName, vPropertyValue);
```
Comment (coder-mike) [inline test/getting-started/code/microvium/microvium.c:4780]: This implementation recurses on the stack. The original implementation did the iteration within the single `while` loop, for stack efficiency.

**Review by coder-mike — CHANGES_REQUESTED** [review]
I'm happy to see the code coverage has gone up! And if ninja speeds up the build/test on Linux, that's great.

I need to think about whether it's suitable to use a property named `__proto__` as the way to store the prototype of an object. My main concern is regarding the performance at runtime when the inheritance chain needs to be iterated. 

BUT, I can see some advantages of doing it that way. For one, it doesn't require the extra 2-byte `dpProto` slot in the `TsPropertyList`, so objects without a prototype will become 2 bytes smaller. But arguably, this is an unusual case.

I'm happy to discuss on Slack. But I'm leaning toward the `dpProto` approach, where the prototype is a special internal field. Similarly, I would probably expect an `ObjectAllocation.proto` or similar in the TypeScript IL, to match this (to keep the engines similar in principle). What do you think?

File: test/getting-started/code/microvium/microvium.c (hunk 4 above)
Comment (pitust) [reply test/getting-started/code/microvium/microvium.c:4780]: AFAIK this does _not_ recurse on the stack, due to tail call optimization (we are, in a sense, JMPing to getProperty).

File: native-vm-just-run-ninja/project.cpp (hunk 3 above)
Comment (pitust) [reply native-vm-just-run-ninja/project.cpp:1]: this is like the windows native-vm-vs-project but on linux, with ninja.

File: native-vm-just-run-ninja/build.ninja (hunk 2 above)
Comment (pitust) [reply native-vm-just-run-ninja/build.ninja:1]: In some tests, whenever cmake is invoked with `-G Ninja` to force ninja as a generator for performance reasons.

File: native-vm-just-run-ninja/.gdb_history (hunk 1 above)
Comment (pitust) [reply native-vm-just-run-ninja/.gdb_history:1]: It shouldn't. My mistake

File: lib/binary-region.ts [hunk 5]
```diff
@@ -53,6 +53,7 @@ export class BinaryRegion {
   public get currentOffset(): Future<number> {
     const address = new Future<number>();
     this.appendSegment(b => {
+      if (address.isResolved) address.unresolve();
```
[the remark this answers is not in the document]
Comment (pitust) [reply lib/binary-region.ts:56]: I was trying to get some weird snapshots (aka. `vmExport(0x12, func); function func() { var x = vmExport; }`) to work. That tripped it up.

File: lib/builtin-globals.ts [hunk 6]
```diff
@@ -5,15 +5,18 @@ import path from 'path';
...
+  const runtimeLib = vm.evaluateModule({ sourceText: runtimeLibText, debugFilename: path.join(__dirname, './runtime-library.mvms') });
 
   const global = vm.globalThis;
   global.Infinity = Infinity;
   global.NaN = NaN;
   global.undefined = undefined;
   const Number = global.Number = vm.newObject();
   Number.isNaN = runtimeLib.Number_isNaN;
-
+  const ContextifiedObject = global.Object = vm.newObject();
```
[the remark this answers is not in the document]
Comment (pitust) [reply lib/builtin-globals.ts:16]: no, you are right.

[the remark this answers is not in the document]
Comment (pitust) [reply lib/builtin-globals.ts:16]: Actually, not the Object prototype, but the global Object constructor itself (not yet a function though). But yes, i will change this

File: native-vm/microvium.c [hunk 7]
```diff
@@ -1261,7 +1261,7 @@ LBL_OP_NUM_OP: {
     MVM_CASE_CONTIGUOUS(VM_NUM_OP_ADD_NUM): {
       CODE_COVERAGE(82); // Hit
       #if MVM_SUPPORT_FLOAT && MVM_PORT_INT32_OVERFLOW_CHECKS
-        #if __has_builtin(__builtin_add_overflow)
+        #if __has_builtin(__builtVM_NUM_OP_ADD_NUMin_add_overflow)
```
[the remark this answers is not in the document]
Comment (pitust) [reply native-vm/microvium.c:1264]: Huh, i should fix that.

File: native-vm-just-run-ninja/build.ninja (hunk 2 above)
Comment (pitust) [reply native-vm-just-run-ninja/build.ninja:1]: Actually i needed to change the invocation of a few things and did that on the side.

File: lib/virtual-machine.ts [hunk 8]
```diff
@@ -1536,9 +1547,6 @@ export class VirtualMachine {
         return;
       }
     } else if (object.type === 'ObjectAllocation') {
-      if (propertyName === '__proto__') {
```
[the remark this answers is not in the document]
Comment (pitust) [reply lib/virtual-machine.ts:1539]: The semantics would be to ignore the prototype.

File: native-vm/microvium.c [hunk 9]
```diff
@@ -3136,11 +3142,11 @@ static Value vm_convertToString(VM* vm, Value value) {
     }
     case TC_REF_FUNCTION: {
       CODE_COVERAGE_UNTESTED(254); // Not hit
-      return VM_NOT_IMPLEMENTED(vm);
+      return mvm_newString(vm, "function () { [microvium code] }", 29);
```
[the remark this answers is not in the document]
Comment (pitust) [reply native-vm/microvium.c:3145]: you are right! I think that mvm_newString could be complemented with mvm_fromCString or something like that.

File: test/getting-started/code/microvium/microvium.c (hunk 4 above)
Comment (coder-mike) [reply test/getting-started/code/microvium/microvium.c:4780]: Ok, yes, it is a tail call, and probably the most common desktop compilers will perform tail call optimization. However, Microvium is optimized for 16-bit embedded platforms, and I'm not sure I can say the same for those. It doesn't seem like tail call optimization is guaranteed as part of the C/C++ spec, as far as I can tell. It's probably safer to just loop.

Comment (pitust) [reply test/getting-started/code/microvium/microvium.c:4780]: It is 1000% safer to loop, especially on an arduino as... 8-bit AVR does *not* handle tail call optimizations apparantly (see https://github.com/wasm3/wasm3/blob/master/docs/Hardware.md, the bottommost target). I should change that and use dpProto.

File: native-vm/microvium.c [hunk 10]
```diff
@@ -3699,11 +3705,6 @@ static TeError getProperty(VM* vm, Value objectValue, Value vPropertyName, Value
   switch (type) {
     case TC_REF_PROPERTY_LIST: {
       CODE_COVERAGE(359); // Hit
-      if (vPropertyName == VM_VALUE_STR_PROTO) {
-        CODE_COVERAGE_UNIMPLEMENTED(326); // Not hit
-        VM_NOT_IMPLEMENTED(vm);
-        return MVM_E_NOT_IMPLEMENTED;
-      }
```
[the remark this answers is not in the document]
Comment (pitust) [reply native-vm/microvium.c:3706]: yeah, that will go back when I have time to implement this.

File: native-vm/microvium.c [hunk 11]
```diff
@@ -3744,9 +3745,22 @@ static TeError getProperty(VM* vm, Value objectValue, Value vPropertyName, Value
...
-      *vPropertyValue = VM_VALUE_UNDEFINED;
-      return MVM_E_SUCCESS;
+      CODE_COVERAGE(326); // Hit
+      if (vPropertyName == VM_VALUE_STR_PROTO) {
+        CODE_COVERAGE(327); // Hit
+        *vPropertyValue = VM_VALUE_NULL;
+        return MVM_E_SUCCESS;
+      }
+      Value proto;
+      TeError e = getProperty(vm, objectValue, VM_VALUE_STR_PROTO, &proto);
```
[the remark this answers is not in the document]
Comment (pitust) [reply native-vm/microvium.c:3755]: It definitly is.

**Pull request closed without merging**